### ISSUE TITLE
Doubt about the updating method of Lagrange Multipliers

### ISSUE THREAD

**Issue by lijie9527** [PKU-Alignment/Safe-Policy-Optimization#69]
from safepo.common.lagrange import Lagrange

nu = 1.0
nu_lr = 0.1
ep_cost = 35
lagrange = Lagrange(
    cost_limit=25.0,
    lagrangian_multiplier_init=1.0,
    lagrangian_multiplier_lr=0.1,
)

print("Before update:")
print(f"Learning Rate: {lagrange.lambda_optimizer.param_groups[0]['lr']}")

lagrange.update_lagrange_multiplier(ep_cost)

learn_lag = lagrange.lagrangian_multiplier

nu += nu_lr * (ep_cost - 25.0)

print(f"Lagrange multiplier: {learn_lag}")
print(f"Nu: {nu}")

There are two methods: one is to treat it as a learnable parameter and use the Adam optimizer, and the other is to directly assign values. Due to the adaptive learning rate of the Adam optimizer, the multipliers obtained by the two are not consistent, and the multiplier changes rapidly in the case of direct assignment under the same learning rate. 

I understand that the original code of PPO Lagrangian uses the former method, while the original code of FOCOPS and CUP seems to use the latter method. Should it be distinguished or can the former be used uniformly?

**Comment by Gaiejj** [comment]
**Why does PPO-Lagrangian employ the Adam optimizer?** This is due to the absence of a universally accepted implementation reference for the PPO-Lagrangian method, and given the oscillatory nature and hyperparameter sensitivity of the Lagrange multipliers, our preliminary experiments indicated that employing the Adam optimizer augments the algorithm's stability and performance.

**Why don't FOCOPS and CUP utilize the Adam optimizer?** This is about the original paper's implementation, as neither the formulation nor the code implementation within the[ FOCOPS](https://github.com/ymzhang01/focops/blob/main/focops_main.py) and [CUP](https://github.com/zmsn-2077/CUP-safe-rl/blob/main/algorithm/cup.py) papers suggested the feasibility of using the Adam optimizer.

As to whether implementing the Adam optimizer within FOCOPS and CUP to enhance algorithmic performance is feasible, we're apprehensive that such a tactic could potentially clash with the original algorithm's convergence rate and direction. While it might yield positive results in some experiments, it's challenging to theoretically validate such effectiveness. Nonetheless, we welcome your attempt and will consider incorporating it as an optional trick in future versions.

**Comment by lijie9527** [comment]
In this case, can I consider that FOCOPS and CUP have no difference in handling cost constraints compared to Lagrangian methods such as PPO-Lagrangian, and their main difference is the way of updating actors?

**Comment by Gaiejj** [comment]
Sure, in code implementation, these three algorithms bear striking similarities. Their difference, indeed, lies solely in the actor-update process.

**Comment by lijie9527** [comment]
The last question is about TRPO class algorithms, such as TRPO, TRPO-Lagrangian, CPO, should they use multiple epochs of full batch or multiple epochs of mini batch to update the critic networks, I found that most of the TRPO class algorithms on the internet use multiple epochs of full batch to update the critic, while most of the PPO class algorithms utilize mini batch. In your implementation, you uniformly use multiple epochs of mini-batch to update the critic, is it because it is more effective and fair to try to preserve the comparison with the first-order methods of the PPO class? 

Also, I found that using multiple epochs of full batch to update the critic of TRPO-like algorithms, the training time is much faster than multiple epochs of mini-batch because the number of updates is much lower, is it possible to adopt multiple epochs of full batch to update the critic in TRPO-based algorithms?

**Comment by Gaiejj** [comment]
In our implementation process, we've referred to [Tianshou](https://github.com/thu-ml/tianshou/blob/master/tianshou/policy/modelfree/trpo.py) and [Stable-Baselines](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/blob/master/sb3_contrib/trpo/trpo.py) and have employed multiple mini-batches for multiple rounds of critic updates. We've experimented with using full-batch updates for the critic in previous tests, but its performance didn't quite measure up to the mini-batch approach.
